0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,262 @@
+9. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the same HMAC key as the sender, derived from the shared secret.
+
+## Data Format
+
+The PSK details are divided into two sets of headers. Each set of headers is formatted like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Application data is appended to the private headers, after a blank line. This object is then encrypted and appended to the public headers as binary data, after a blank line. Both private and public headers are parsed in the exact same way as HTTP headers.
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Nonce: fpwpAhlN588
+Key: hmac-sha-256
+Encryption: aes-256-ctr
+Header: data_everyone_can_see
+
+...
+```
+
+The public headers come after a single status line, which must be `PSK/1.0`. Public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+PSK payments use an HMAC of the PSK details as a fulfillment, to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The decryption key is derived from the pre-shared secret key, and the AES-256-CTR initialization vector is set to the value of the `Nonce` header (`fpwpAhlN588` in the above example). The nonce MUST be generated with cryptographically-secure randomness. **If the nonce is reused with the same shared secret, it could leak unencrypted data or allow money to be stolen by malicious parties.**
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ The HMAC algorithm, which MUST be `hmac-sha-256` (case sensitive) |
+| `Nonce` | _(Required)_ A [base64url-encoded](https://en.wikipedia.org/wiki/Base64#URL_applications) nonce, used to generate ensure uniqueness of the PSK data and as an initialization vector (IV) for AES-256-CTR encryption. The nonce **MUST** be 16 bytes, and **MUST** be generated with strong randomness. |
+| `Encryption` | _(Required)_ MUST be set to `aes-256-ctr` or `none`. If it is set to `aes-256-ctr`, then private headers and application data will be AES-256-CTR encrypted. If it is set to `none`, then private headers and application data will be in cleartext. This cleartext will still be appended to the public headers after an empty line. |
```

Comment:
Same as the above comment, I think we should say that PSK 1.0 implementations must at minimum support this algorithm. If two implementations support some other algorithm and want to use it that's fine.

That made me think: we might want to define standard errors for PSK. If you use a `Key` or `Encryption` algorithm that isn't supported, it would be helpful if it would tell you in the same way each time.